File: qviability/core/measure.py
"""Thresholding and measurement.

All math runs on the raw pixel array. ``positive`` pixels are those at or
above the threshold (``raw >= threshold``) — matching ImageJ's "dark"
auto-threshold, which selects the bright object on a dark background.

Exclusion boxes (e.g. drawn over a scale bar) remove pixels from both the
positive count *and* the total, so they never affect any measurement.
"""

from __future__ import annotations

import cv2
import numpy as np

from .project import Measurement

# Overlay tint per channel (R, G, B), drawn translucently over positive pixels.
OVERLAY_COLORS = {
    "Green": (40, 255, 70),     # green tint on green/GFP images
    "Red": (255, 40, 220),      # magenta on red/RFP images (visible on red)
}
OVERLAY_ALPHA = 115             # ~45 % opacity
EXCLUDE_RGBA = (140, 140, 140, 160)   # gray fill over excluded regions

# Bright opaque ring drawn just outside marked regions when the tuning-screen
# "highlight marks" aid is on — makes tiny blobs pop. Display only; never
# exported and never affects measurement.
OUTLINE_RGBA = (255, 230, 0, 255)     # high-visibility yellow ring


def apply_threshold(raw: np.ndarray, threshold: int) -> np.ndarray:
    """Boolean mask of positive (>= threshold) pixels."""
    return raw >= threshold


def exclusion_mask(shape, exclusions) -> np.ndarray | None:
    """Build a bool mask (True = excluded) from a list of [x, y, w, h] rects.

    Returns None when there are no exclusions so callers can keep a fast path.
    Rectangles are clamped to the image bounds.
    """
    if not exclusions:
        return None
    h, w = shape[:2]
    mask = np.zeros((h, w), dtype=bool)
    for rect in exclusions:
        x, y, rw, rh = (int(round(v)) for v in rect)
        x0, y0 = max(0, x), max(0, y)
        x1, y1 = min(w, x + rw), min(h, y + rh)
        if x1 > x0 and y1 > y0:
            mask[y0:y1, x0:x1] = True
    return mask


def measure(raw: np.ndarray, threshold: int, method: str = "Manual",
            exclude_mask: np.ndarray | None = None) -> Measurement:
    """Measure the positive region of *raw* at *threshold*.

    Computes all four candidate metrics so the ambiguous "intensity" choice
    can be made later in Excel:
      * positive_pixels       — count of pixels >= threshold
      * positive_area_pct     — that count as a percentage of the image
      * mean_intensity_positive — mean raw value of positive pixels
      * integrated_intensity  — sum of raw values of positive pixels (ImageJ RawIntDen)

    Pixels under *exclude_mask* are dropped from both the positive count and
    the total, so the percentage stays correct.
    """
    if exclude_mask is not None:
        valid = ~exclude_mask
        positive = (raw >= threshold) & valid
        total = int(valid.sum())
    else:
        positive = raw >= threshold
        total = int(raw.size)

    count = int(positive.sum())
    if count > 0:
        vals = raw[positive].astype(np.float64)
        mean_intensity = float(vals.mean())
        integrated = float(vals.sum())
    else:
        mean_intensity = 0.0
        integrated = 0.0

    return Measurement(
        threshold=int(threshold),
        positive_pixels=count,
        total_pixels=total,
        positive_area_pct=(count / total * 100.0) if total else 0.0,
        mean_intensity_positive=mean_intensity,
        integrated_intensity=integrated,
        method=method,
    )


def _outline_mask(mask: np.ndarray, thickness: int = 2) -> np.ndarray:
    """Bool ring just outside *mask* — its dilation minus itself."""
    m = mask.astype(np.uint8)
    kernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (3, 3))
    dilated = cv2.dilate(m, kernel, iterations=thickness)
    return (dilated > 0) & ~mask


def overlay_rgba(positive: np.ndarray, channel: str,
                 exclude_mask: np.ndarray | None = None,
                 outline: bool = False) -> np.ndarray:
    """Build an HxWx4 uint8 RGBA overlay.

    Positive pixels get the channel tint; excluded pixels get a gray fill
    (and are never shown as positive, since they are not counted).

    When *outline* is True a bright opaque ring is drawn just outside each
    marked region — a tuning-screen aid for spotting small blobs. It is not
    used for export and does not affect any measurement.
    """
    color = OVERLAY_COLORS.get(channel, OVERLAY_COLORS["Green"])
    h, w = positive.shape
    rgba = np.zeros((h, w, 4), dtype=np.uint8)

    counted = positive & ~exclude_mask if exclude_mask is not None else positive
    rgba[counted] = (color[0], color[1], color[2], OVERLAY_ALPHA)
    if outline and counted.any():
        rgba[_outline_mask(counted)] = OUTLINE_RGBA
    if exclude_mask is not None:
        rgba[exclude_mask] = EXCLUDE_RGBA
    return rgba

File: web/src/core/measure.ts
// 1:1 port of qviability/core/measure.py.
//
// All math runs on the raw pixel array. `positive` pixels are those at or above
// the threshold (raw >= threshold) — matching ImageJ's "dark" auto-threshold.
// Exclusion boxes remove pixels from BOTH the positive count and the total, so
// they never affect any measurement.

import type { Measurement, Rect } from "./types";

// Overlay tint per channel (R, G, B), drawn translucently over positive pixels.
export const OVERLAY_COLORS: Record<string, [number, number, number]> = {
  Green: [40, 255, 70],
  Red: [255, 40, 220],
};
export const OVERLAY_ALPHA = 115; // ~45% opacity
export const EXCLUDE_RGBA: [number, number, number, number] = [140, 140, 140, 160];
export const OUTLINE_RGBA: [number, number, number, number] = [255, 230, 0, 255];

export type Mask = Uint8Array; // 1 = set, 0 = clear

/** Boolean (0/1) mask of positive (>= threshold) pixels. */
export function applyThreshold(
  raw: Uint8Array | Uint16Array,
  threshold: number,
): Mask {
  const out = new Uint8Array(raw.length);
  for (let i = 0; i < raw.length; i++) out[i] = raw[i] >= threshold ? 1 : 0;
  return out;
}

/**
 * Build a mask (1 = excluded) from a list of [x, y, w, h] rects.
 * Returns null when there are no exclusions so callers can keep a fast path.
 * Rectangles are clamped to the image bounds.
 */
export function exclusionMask(
  width: number,
  height: number,
  exclusions: Rect[],
): Mask | null {
  if (!exclusions || exclusions.length === 0) return null;
  const mask = new Uint8Array(width * height);
  for (const rect of exclusions) {
    const x = Math.round(rect[0]);
    const y = Math.round(rect[1]);
    const rw = Math.round(rect[2]);
    const rh = Math.round(rect[3]);
    const x0 = Math.max(0, x);
    const y0 = Math.max(0, y);
    const x1 = Math.min(width, x + rw);
    const y1 = Math.min(height, y + rh);
    for (let yy = y0; yy < y1; yy++) {
      const row = yy * width;
      for (let xx = x0; xx < x1; xx++) mask[row + xx] = 1;
    }
  }
  return mask;
}

/**
 * Measure the positive region of `raw` at `threshold`.
 * Pixels under `excludeMask` are dropped from both the positive count and the
 * total, so the percentage stays correct.
 */
export function measure(
  raw: Uint8Array | Uint16Array,
  threshold: number,
  method = "Manual",
  excludeMask: Mask | null = null,
): Measurement {
  let count = 0;
  let total: number;
  let sum = 0;

  if (excludeMask) {
    let valid = 0;
    for (let i = 0; i < raw.length; i++) {
      if (excludeMask[i]) continue;
      valid++;
      if (raw[i] >= threshold) {
        count++;
        sum += raw[i];
      }
    }
    total = valid;
  } else {
    total = raw.length;
    for (let i = 0; i < raw.length; i++) {
      if (raw[i] >= threshold) {
        count++;
        sum += raw[i];
      }
    }
  }

  const mean_intensity = count > 0 ? sum / count : 0.0;
  const integrated = count > 0 ? sum : 0.0;

  return {
    threshold: Math.trunc(threshold),
    positive_pixels: count,
    total_pixels: total,
    positive_area_pct: total ? (count / total) * 100.0 : 0.0,
    mean_intensity_positive: mean_intensity,
    integrated_intensity: integrated,
    method,
  };
}

/** Bool ring just outside `mask` — its dilation minus itself (display only). */
export function outlineMask(
  positive: Mask,
  width: number,
  height: number,
  thickness = 2,
): Mask {
  // cv2.getStructuringElement(MORPH_ELLIPSE, (3,3)) is the 4-connected cross;
  // dilating `thickness` times grows a diamond of that radius.
  let cur = positive;
  for (let it = 0; it < thickness; it++) {
    const next = new Uint8Array(width * height);
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        const i = y * width + x;
        if (
          cur[i] ||
          (x > 0 && cur[i - 1]) ||
          (x < width - 1 && cur[i + 1]) ||
          (y > 0 && cur[i - width]) ||
          (y < height - 1 && cur[i + width])
        ) {
          next[i] = 1;
        }
      }
    }
    cur = next;
  }
  const ring = new Uint8Array(width * height);
  for (let i = 0; i < ring.length; i++) ring[i] = cur[i] && !positive[i] ? 1 : 0;
  return ring;
}

/**
 * Build an HxWx4 uint8 RGBA overlay (row-major, RGBA per pixel).
 * Positive pixels get the channel tint; excluded pixels get a gray fill.
 */
export function overlayRgba(
  positive: Mask,
  width: number,
  height: number,
  channel: string,
  excludeMask: Mask | null = null,
  outline = false,
): Uint8ClampedArray {
  const color = OVERLAY_COLORS[channel] ?? OVERLAY_COLORS.Green;
  const rgba = new Uint8ClampedArray(width * height * 4);

  const counted = new Uint8Array(width * height);
  for (let i = 0; i < counted.length; i++) {
    counted[i] = positive[i] && !(excludeMask && excludeMask[i]) ? 1 : 0;
  }

  for (let i = 0; i < counted.length; i++) {
    if (counted[i]) {
      const o = i * 4;
      rgba[o] = color[0];
      rgba[o + 1] = color[1];
      rgba[o + 2] = color[2];
      rgba[o + 3] = OVERLAY_ALPHA;
    }
  }

  if (outline) {
    const ring = outlineMask(counted, width, height);
    for (let i = 0; i < ring.length; i++) {
      if (ring[i]) {
        const o = i * 4;
        rgba[o] = OUTLINE_RGBA[0];
        rgba[o + 1] = OUTLINE_RGBA[1];
        rgba[o + 2] = OUTLINE_RGBA[2];
        rgba[o + 3] = OUTLINE_RGBA[3];
      }
    }
  }

  if (excludeMask) {
    for (let i = 0; i < excludeMask.length; i++) {
      if (excludeMask[i]) {
        const o = i * 4;
        rgba[o] = EXCLUDE_RGBA[0];
        rgba[o + 1] = EXCLUDE_RGBA[1];
        rgba[o + 2] = EXCLUDE_RGBA[2];
        rgba[o + 3] = EXCLUDE_RGBA[3];
      }
    }
  }

  return rgba;
}
